# nextflow_workflow
I have written countless bioinformatics pipelines over the years, mostly using Bash scripting to organize program and alanysis flow (using programs/packages/libraries that I, or others, have written in C/C++, Perl,R,Python,Bash itself, or even MATLAB). These pipelines that I would generally organize to submit on various job scheduling software (PBS/Torque, Slurm, etc.) on HPC (High perfomance computing) clusters.
I have found, recently, that Nextflow offers a well-organized, easy-to-use approach to crafting pipelines and employing containerized software dependencies (obviating many of the problems associated with versioning issues that often plague the bioinformatics and computational biology worlds).

Here we have an example where I (as many other computational biologists and bioinformaticians) have organized their numerous scripts and programs over the years according to function or project. 

![image](https://github.com/user-attachments/assets/c2b1e458-46e7-4bb7-b310-27f16b67860b)

Using the Nextflow architecture, we can create and organize workflows based on these scripts to enhance code reusability and portability (e.g. among collaboraters or across systems)
Here I present my own end-to-end RNA-Seq analysis Nextflow workflow, using my own scripts that I had created for various previous (bulk) RNA-Seq analyses, and using the publicly available Mus musculus RNA-Seq data (.fastq format) from the now-classic "A Beginnerâ€™s Guide to Analysis of RNA Sequencing Data" (PMID: 29624415) - USA, National Institutes of Health (NIH), National Library of Medicine (NLM) Accession 
PRJNA450151 (https://www.ncbi.nlm.nih.gov/bioproject/PRJNA450151)

![nextfoww_example_workflow](https://github.com/user-attachments/assets/01c00510-54d2-462c-9599-b37d5c8e3940)

Here we have my Nextflow workflow (file: nextflow_workflow.nf) version running on the Ubuntu App on Windows 10 with Windows Subsystem for Linux 2 (WSL2) installed and enabled. on an ASUS PC
![image](https://github.com/user-attachments/assets/aa8d2539-29ca-4a72-a1a6-c13fc3fc8960)

Addiitonally, one can use Gitpod to develop and test their nextflow project as well as link it to a respective GitHub project page, in this case: https://github.com/mmariani123/nextflow_workflow . Changes to the Nextflow files (.nf), for example can then be pushed to GitHub

![gitpod_commit_1](https://github.com/user-attachments/assets/0b8fab31-85cc-409b-b84a-f0fc66c7548c)

One can also use the file explorer in Gitpod to navigate through files generated by our nextflow pipeline such as the quality control output from FASTQC. But it is important to note that the free tier of Github only allows for a maximum of 8GB of RAM and 30GB of storage, quite low for today's bioinformatics needs. Additionally, GuitHub itself has a maximum file size of 100MB; otherwise, GitHub's LTS system will need to be employed. Ideally, nextflow would be run on an HPC cluster (either physical or cloud) with adequate computational resources. 

![image](https://github.com/user-attachments/assets/fe2e1e9f-dcd2-4a2f-b40a-42c4a602bdee)


